### PR TITLE
feat(router): add support for hash-based location

### LINKED ISSUE #2555: feat(router): support hashchange mode

### PR DESCRIPTION
**Pull request by btford**
Closes #2555

### CODE REVIEW

**Comment by btford** [comment]
I'm not sure how to test this (except via an e2e test). Any advice?

File: modules/angular2/src/router/hash_location_strategy.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import {DOM} from 'angular2/src/dom/dom_adapter';
+import {Injectable} from 'angular2/di';
+import {LocationStrategy} from './location_strategy'
+import {EventListener, History, Location} from 'angular2/src/facade/browser';
+
+@Injectable()
+export class HashLocationStrategy extends LocationStrategy {
```
Comment (rkirov) [inline modules/angular2/src/router/hash_location_strategy.ts]: I notice it is almost identical with Html5LocationStrategy, except one '#'. But it is simple enough to keep as a separate class for now, if things start to get more complication you can explore factoring the commonalities into a base class.

**Comment by rkirov** [comment]
It is not testable, because our DOM adapter is not injectable. I dont' see an easy wait to fix that, without changing that.

File: modules/angular2/src/router/hash_location_strategy.ts (hunk 1 above)
Comment (btford) [inline modules/angular2/src/router/hash_location_strategy.ts]: Depending on browser weirdness, I might have to do some shimming WRT history and hashchanges. I think/hope that evergreen browsers do the right thing, but historically that has not been true. So the behavior here in each class might get a bit more complicated. If not, I'll definitely revisit your suggestion :+1:

**Comment by btford** [comment]
Landed as a67f2314f997a785d01163498b231435cb9b4680.

**Comment by dsebastien** [comment]
works as expected, thanks!

**Pull request merged**